bindings/python/iree/compiler/debugging.py

```diff
@@ -0,0 +1,172 @@
+
+  The outer-most creator can customize it with explicit arguments to __init__
+  but these will be ignored if an instance is already thread bound.
+  """
+  TEMP_PATH_ENV_KEY = "IREE_SAVE_TEMPS"
+
+  @staticmethod
+  def implicit():
+    stack = _get_temp_file_saver_stack()
+    if stack:
+      return stack[-1]
+    return TempFileSaver()
+
+  def __init__(self,
+               temp_path_pattern: str = None,
+               *,
+               invocation_id: Optional[str] = None):
+    self.retained = False
+    self._refcount = 0
+    if temp_path_pattern is None:
+      temp_path_pattern = os.environ.get(TempFileSaver.TEMP_PATH_ENV_KEY)
+    if temp_path_pattern is None:
+      return
+
+    global _invocation_id
+    if invocation_id is not None:
+      self.invocation_id = invocation_id
+    else:
+      self.invocation_id = _invocation_id
+      _invocation_id += 1
+
+    self.retained_path = _interpolate_path_pattern(
+        temp_path_pattern, invocation_id=self.invocation_id)
+    self.retained = True
+    self._retained_file_names = set()
+    self._copy_on_finalize = list()  # Of (source_path, target_path)
+    self._created_dir = False
+
+  def __enter__(self):
+    _get_temp_file_saver_stack().append(self)
+    self._refcount += 1
+    return self
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    del _get_temp_file_saver_stack()[-1]
+    self._refcount -= 1
+    if self._refcount == 0:
+      self._finalize()
+
+  @staticmethod
+  def current():
+    try:
+      return _get_temp_file_saver_stack()[0]
+    except KeyError:
+      raise RuntimeError("No current TempFileSaver")
+
+  def alloc_optional(self,
+                     file_name: str,
+                     *,
+                     export_as: Optional[str] = None) -> Optional[str]:
+    """Allocates an optional temporary file.
+
+
+    When in non-retained mode, the return value is 'export_as', meaning that the
+    file is just some user specified output file.
+
+    When in retained mode, the output file will be an index-mangled variant
+    of 'file_name' under the temp_path. In addition, a mapping will be added
+    so that upon finalization, the file is also exported to 'export_as' if
+    specified.
+
+    Returns None if neither a user-specified 'export_as' is specified nor in
+    retained mode.
+    """
+    if not self.retained:
+      return export_as
+    alloced_path = self._alloc_retained_path(file_name)
+    if export_as:
+      self._copy_on_finalize.append((alloced_path, export_as))
+    return alloced_path
+
+  def _alloc_retained_path(self, file_name: str) -> str:
+    assert self.retained
+    index = 0
+    original_file_name = file_name
+    while True:
+      if file_name not in self._retained_file_names:
+        # First use of this name.
+        self._retained_file_names.add(file_name)
+        if not self._created_dir:
+          os.makedirs(self.retained_path, exist_ok=True)
+          self._created_dir = True
```

Comment:
Is the explicit check here actually singificantly cheaper than calling `os.makedirs` with `exist_ok=True`?

Reply:
Convinced. Simplified.